Http Refactor

jochem725 opened the pull request:
This branch provides the HTTP Refactor. This should be merged in mjpeg-stream-distribute, from which it origins.

Please verify carefully if the API is still acting the same.
Everything now first passes a LogHandler, which logs the URI. In the future maybe this handler can also do the URI verification.

Kingdorian commented on backend/.checkstyle:
```diff
@@ -1,7 +1,7 @@
 
 <fileset-config file-format-version="1.2.0" simple-config="false" sync-formatter="false">
-  <local-check-config name="maven-checkstyle-plugin verify-style" location="file:/D:/Git/Contextproject-BeNine/checkstyle.xml" type="remote" description="maven-checkstyle-plugin configuration verify-style">
+  <local-check-config name="maven-checkstyle-plugin verify-style" location="file:/C:/Users/bryan/Documents/Git/Contextproject-BeNine/checkstyle.xml" type="remote" description="maven-checkstyle-plugin configuration verify-style">
```
Shouldn't this be a more generic name instead of the path that obviously only works on Bryan's machine?

bvanwijk replied:
This file is removed from git in https://github.com/Kingdorian/Contextproject-BeNine/pull/122/commits/956ca842581d7ae6b751f9cd84299bae10317601

NaomidR replied:
Checkstyle is already added to gitignore as well.

Kingdorian commented on backend/src/main/java/com/benine/backend/http/CameraFocusHandler.java:
```diff
@@ -0,0 +1,50 @@
+import com.benine.backend.camera.FocussingCamera;
+import org.eclipse.jetty.server.Request;
+
+import java.io.IOException;
```
I see jetty and javafx imports for http things should these two be mixed together?

jochem725 replied:
They are not JavaFX imports, but javax imports. This should be correct.

Kingdorian replied:
Whoops you're right, it should be fined then.

Kingdorian commented on backend/src/main/java/com/benine/backend/http/CameraInfoHandler.java:
```diff
@@ -1,24 +1,124 @@
+  /**
+   * Constructor for a new CameraInfoHandler, handling the /camera/ request.
+   */
+  public CameraInfoHandler() {
```
Wouldn't it be better to just have a list of handlers?

Kingdorian commented on backend/src/main/java/com/benine/backend/http/CameraInfoHandler.java:
```diff
@@ -1,24 +1,124 @@
+      String route = getRoute(request);
+      Camera cam = getCameraController().getCameraById(getCameraId(request));
+
+      switch (route) {
```
I see lots of duplicate code here. Wouldn't it be better to store a string in each handler, create a local var handler and then loop over the list of handlers setting the local var handler to the handler with the string matching to route and then handle the request with that handler.

Kingdorian commented on backend/src/main/java/com/benine/backend/http/CameraRequestHandler.java:
```diff
@@ -0,0 +1,42 @@
+   */
+  public int getCameraId(Request request) {
+    // Get path
+    Pattern pattern = Pattern.compile(".*/(\\d*)/.*");
```
I think this also accepts wrong strings like /preset/1/camera/4/, I would change this to something like .*/camera/(\\d*)//.

Kingdorian commented on backend/src/main/java/com/benine/backend/http/CameraRequestHandler.java:
```diff
@@ -0,0 +1,42 @@
+
+    Matcher m = pattern.matcher(path);
+
+    if (m.matches()) {
```
You can clean this up by doing something like "return m.matches ? Intteger.parseInt(m.group(1) : -1"

Kingdorian commented on backend/src/main/java/com/benine/backend/http/CameraStreamHandler.java:
```diff
@@ -0,0 +1,88 @@
+    try {
+      streamReader = ServerController.getInstance().getStreamController().getStreamReader(camID);
+    } catch (StreamNotAvailableException e) {
+      e.printStackTrace();
```
Please log an exception here like "getLogger.log([String:description], e);"  this also automatically logs the stacktrace (and is thus stored in the logs).

Kingdorian commented on backend/src/main/java/com/benine/backend/http/CameraStreamHandler.java:
```diff
@@ -0,0 +1,88 @@
+   * @param httpServletResponse   The response for which the headers should be set.
+   */
+  private void setHeaders(MJPEGStreamReader reader, HttpServletResponse httpServletResponse) {
+    httpServletResponse.setContentType("multipart/x-mixed-replace;boundary="
```
I don't really like the idea of hardcoding this. Maybe you can make a simple file like with the default headers for example which then can be read by a simple BufferedReader.

jochem725 replied:
I think storing it in a file requires it to be read every time. Maybe a better solution would be to create an ENUM or final variables for these headers?

Kingdorian replied:
Well you would only read the file on server creation but a constant would be fine too.

Kingdorian commented on backend/src/main/java/com/benine/backend/http/CameraZoomHandler.java:
```diff
@@ -0,0 +1,62 @@
+    String zoomType = request.getParameter("zoomType");
+    String zoom = request.getParameter("zoom");
+
+    System.out.println(zoomType);
```
I think we've got a logger for this.

jochem725 replied:
This was some small debug code I forgot to remove. Thanks for pointing it out.

Kingdorian commented on backend/src/main/java/com/benine/backend/http/PresetsCreatePresetHandler.java:
```diff
@@ -0,0 +1,135 @@
+
+      if (camera instanceof IPCamera) {
+        IPCamera ipcam = (IPCamera) camera;
+        System.out.println(ipcam);
```
System.out again? I think it would be better to either remove this or change this to the logger.

Kingdorian commented on backend/src/main/java/com/benine/backend/http/PresetsHandler.java:
```diff
@@ -0,0 +1,108 @@
+   * Constructor for a new PresetsHandler, handling the /presets/ request.
+   */
+  public PresetsHandler() {
+    this.createPreset = new PresetsCreatePresetHandler();
```
Same comment as for the CameraHandler.

Kingdorian commented on backend/src/main/java/com/benine/backend/http/PresetsRecallPresetHandler.java:
```diff
@@ -0,0 +1,76 @@
+   * @throws CameraConnectionException  If the camera cannot be reached.
+   * @throws MalformedURIException      If the request contains an error.
+   */
+  public void moveCamera(Camera camera, Preset preset)
```
Maybe it would be better to create a preset.apply(Camera c) function but that might be out of scope for this pullrequest.

jochem725 replied:
I do agree with you. I think moving the camera itself should not be a responsibility for the http endpoint. Maybe we should create an issue for this.

Kingdorian commented on backend/src/test/java/com/benine/backend/http/CameraStreamHandlerTest.java:
```diff
@@ -0,0 +1,112 @@
+      e.printStackTrace();
+    }
+  }
+
```
Could the @Test annotation be missing?

jochem725 replied:
No this is correct, since a thread is needed in most of the tests this is a separate method creating the thread.

Kingdorian commented:
Looks good in general but there definitely are some things that need to be addressed in my opinion. Also try to watch out for deleting and replacing files manually next time, this creates a lot of unnecessary things to be marked as changed by github which definitely does not make things more clear.

jochem725 commented:
@Kingdorian I've addressed all your comments. Something with the files went wrong in my IDE, but I was unable to fix it. I'll be more careful next time. Of course things can be improved such as the way requests are validated now. This can be done in a separate handler for instance.

bvanwijk commented:
Looks good, however there are some checkstyle warnings:
![2016-05-24](https://cloud.githubusercontent.com/assets/9803346/15504756/cb110b58-21c0-11e6-9373-81cbc7b82ab3.png)

If these are fixed I agree with merging.

NaomidR commented on backend/src/main/java/com/benine/backend/http/CameraFocusHandler.java:
```diff
@@ -0,0 +1,56 @@
+public class CameraFocusHandler extends CameraRequestHandler {
+
+  @Override
+  public void handle(String s, Request request, HttpServletRequest req, HttpServletResponse res)
```
The handle method seems to be very long. Even though it is clear what is done here, maybe this method can be split into smaller methods?

NaomidR commented on backend/src/main/java/com/benine/backend/http/CameraIrisHandler.java:
```diff
@@ -0,0 +1,56 @@
+public class CameraIrisHandler extends CameraRequestHandler {
+
+  @Override
+  public void handle(String s, Request request, HttpServletRequest req, HttpServletResponse res)
```
Same here, it is pretty long. It doesn't really bother me but I want to have your opinion on why you didn't split this method.

NaomidR commented on backend/src/main/java/com/benine/backend/http/PresetsRecallPresetHandler.java:
```diff
@@ -0,0 +1,76 @@
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+public class PresetsRecallPresetHandler extends RequestHandler {
```
I think that the name of this class is a bit vague. Can this maybe be changed to something that's more clear?

jochem725 replied:
Presets is a prefix for the route it is executed on. Maybe I can make this more clear by renaming and using inheritance.

Kingdorian commented:
I think you can merge this.
